---
title: "Differential Expression Analysis"
author: "Mikhail Dozmorov"
date: "October 25, 2014"
output: html_document
---
Workshop at Rajiv Gandhi Center for Biotechnology

January 16-18, 2013

Material is public domain
```{r setup, echo=FALSE}
# Set up the environment
library(knitr) 
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=T, tidy=T, fig.keep='high', echo=T, dpi=300, out.width=700)
options(replace.assign=TRUE, width=220)
suppressMessages(library(pander))
panderOptions('table.split.table', Inf)
set.seed(1)
```


Limma
===
Limma is a software package for the analysis of gene expression microarray data, especially the use of linear models for analyzing designed experiments and the assessment of differential expression. The package includes pre-processing capabilities for two-color spotted arrays. The differential expression methods apply to all array platforms and treat Affymetrix, single channel and two channel experiments in a unified way. The methods are described in [Smyth 2004] (https://www.ncbi.nlm.nih.gov/pubmed/15297296) and in the [limma manual](http://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf). An illustrated introduction for the GUI packages can be found at Walter+Eliza Hall Bioinformatics Institute of Medical Research [(WEHI)](http://bioinf.wehi.edu.au/limma/index.html).

We will be analyzing tissue-specific differences from [Su et al., 2002](http://www.ncbi.nlm.nih.gov/pubmed/11904358). Prepare the data:
```{r, warning=FALSE}
source("http://bioconductor.org/biocLite.R") # Import biocLite() function into R environment
# biocLite("limma")
library(limma)
# limmaUsersGuide() # Opens pdf manual for limma
suppressMessages(library(affy))
eset.rma <- justRMA(celfile.path="data/Su_CELs/") # RMA summarization of the CEL files
pData(eset.rma) # Check what samples we have
```

There are two different ways to form the design matrix. We can either
1. create a design matrix which includes a coefficient for the treated vs wild type difference, or
2. create a design matrix which includes separate coefficients for wild type and mutant mice and then extract the difference as a contrast.

For the first approach, the treatment-contrasts parametrization, the design matrix should be as follows:
```{r}
# Design matrix: Treatment-constrast parametrization
a<-rep(0,length(pData(eset.rma)$sample)) # Create a vector of 0
a[grep("liver", rownames(pData(eset.rma)), ignore.case=T)] <-1 # Mark "liver" conditions as "1"
a # Check your work
design <- cbind(Brain=1, LiverVsBrain=a) # Columnwise bind
design # Check your work
```

Here, the first coefficient estimates the mean log-expression for brain tissue and plays a role of an intercept. The second coefficient estimates the difference between brain and liver cells. Differentially expressed genes can be found by:
```{r}
# Identifying differentially expressed genes
fit <- lmFit(eset.rma, design)
fit <- eBayes(fit)
result <- topTable(fit, number=10, sort.by="B", adjust="BH", p.value=0.05) # Get top 20 differentially expressed genes
pander(result) # Check your work
```

For the second approach, the group-means parametrization, the design matrix can be computed by:
```{r}
# Design matrix: separate group coefficients
design <- cbind(Brain=c(rep(1,4),rep(0,4)), 
                Liver=c(rep(0,4),rep(1,4))) # Manually create design matrix
design # Check
design <- model.matrix(~0+factor(a)) # Another way - factor makes two levels, one for each group
colnames(design) <- c("Brain", "Liver") # Label columns properly
design # Check your work
```

To find differentially expressed genes, group-means parametrization should be converted into contrast matrix:
```{r}
fit <- lmFit(eset.rma, design)
cont.matrix <- makeContrasts(BrainvsLiver=Brain-Liver, levels=design) # Make matrix of contrasts
cont.matrix # See what's inside
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2)
pander(topTable(fit2, adjust="BH"))
```

The above approaches for two groups extend easily to any number of groups. Suppose that we want to pairwise compare all four conditions. An appropriate design matrix can be created using:
```{r}
# Several groups
a <- c(1, 1, 2, 2, 3, 3, 4, 4) # Four conditions, two replicates per condition
design <- model.matrix(~0+factor(a)) # Now we have four levels for design matrix
colnames(design) <- c("B", "fB", "fL", "L") # label columns
design # Check your work
```

We create contrast matrix for three pairwise comparisons, for the sake of visualizing the results in a form of Venn diagram (3 comparisons at a time). Finding differentially expressed genes are the same:
```{r}
contrast.matrix <- makeContrasts(B-fB, L-fL, B-L, levels=design) # Make three contrasts
contrast.matrix # Check your work
fit <- lmFit(eset.rma, design)
fit2 <- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)
result <- decideTests(fit2, adjust="BH", p.value=0.05, lfc=log2(2))
```

Use decideTests function to have a summary of the results for Venn diagram, and visualize it:
```{r}
vennDiagram(result) # How genes differentially expressed in different conditions
vennDiagram(result,include="up") # Only upregulated
vennDiagram(result,include="down") # Or downregulated
```

We can save our results into a file:
```{r}
write.table(topTable(fit2,coef=1,number=1000,adjust.method="BH",p.value=0.05,lfc=log2(2)),"results/filename.txt",sep="\t") # vary coefficient to write corresponding results to a tab-separated file
```

SAM (Significance Analysis of Microarrays)
===
```{r, warning=FALSE}
# biocLite("samr")
library(samr) # Load the library
library(limma) # To get access to normalizeQuantiles
library(affy) # Load affy package 
library(GEOquery)
# Download GDS file, put it in the current directory, and load it:
# gds858 <- getGEO('GDS858', destdir="data") 
gds858<-getGEO(filename="data/GDS858.soft.gz") # If FTP doesn't work, read in from local file
eset <- GDS2eSet(gds858, do.log2=TRUE) # Convert the data to ESET object
# help(ExpressionSet) # If needed, refresh your memory
```

Let's check which infection statuses do we have:
```{r}
pData(eset)$infection # Let's check the infection status
table(pData(eset)$infection) # How many samples are in each infection status
```

Select indexes of the columns associated with "uninfected" or "FRD440" infection status. Check that the right infection status was selected:
```{r}
selected<-grep("uninfected|FRD440",pData(eset)$infection) # Select indexes
pData(eset)$infection[selected] # Check if the right infection status was selected
```

We selected 4 "uninfected" samples and 4 "FRD440" samples. Let's make a vector of outcome measurements, defining groups of replicates. "Uninfected" samples are labeled as "1" and "FRD440" samples are labeled as "2":
```{r}
y=c(rep(1,4),rep(2,4)) # Vector of outcome measurements
y # Visualize it
```

From the whole expression matrix we select a subset of columns corresponding to the selected indexes. We then quantile normalize the data:
```{r}
exprs.selected<-exprs(eset)[,selected] # Get a subset of expression values of the selected samples
exprs.selected.q<-normalizeQuantiles(exprs.selected) # Quantile normalize the data
```

In order to perform the analysis, SAM requires a list object containing a data object with expression values in a form of p by n matrix, one observation per column (missing values allowed),  a vector of length n of outcome measurements, vectors of gene names and gene IDs, both of length p, and a boolean indicating whether the data is log2-transformed. This object resembles ExpressionSet with slots having different data, but we assemble it as a list.

Now we have our expression matrix (exprs.selected.q), our vector of outcomes (y). Let's use row names of the expression matrix as both gene names and IDs, and assemble our data:
```{r}
genenames<-rownames(exprs(eset)) # Get row names = gene names or IDs
data<-list(x=exprs.selected.q,y=y, geneid=genenames,genenames=genenames,logged2=T)
```

SAM can perform a variety types of analyses, specified in "resp.type" parameter. Each analysis type requires specific formatting of outcome vector and expression data. Refer to help(samr) for the details. Now, we're performing "Two class unpaired" analysis:
```{r, results='hide'}
samr.obj<-samr(data,resp.type="Two class unpaired",nperms=100)
```

Let's check what we have. Everything obvious, isn't it?
```{r}
names(samr.obj) # Get object names from samr.obj
```

Now, we have to choose the delta value that is able to give the best compromise in terms of called genes, false genes and False Discovery Rate (FDR). In microarray analysis is very important to have statistically robust results, but we have keep in mind that too small sized results are not able to describe the biological meaning of the experiment. In any case, keeping the FDR < 10% (the number of false positives is < 10%) is pretty safe in most of the cases.

In general, defining the cut-off is a subjective choice and there is no absolute best way to do it.
```{r, results='hide'}
delta.table<-samr.compute.delta.table(samr.obj, min.foldchange=1.5) # Compute thresholds for different deltas
pander(delta.table) # Look at the whole range
```

Let's select delta with median FDR <10% - subset the whole delta table and take the first row.
```{r}
delta.table[delta.table[,"median FDR"] < 0.1,][1,] # Check delta corresponding to median FDR ~0.1
delta<-1.5 # Select the delta
```

Let's select the delta, and have a look at SAM plot:
```{r}
samr.plot(samr.obj,delta) # Check SAM plot
```

Do we have larger number of upregulated genes? Or downregulated? Let's have a look at them:
```{r}
siggenes.table<-samr.compute.siggenes.table(samr.obj,delta,data,delta.table,min.foldchange=1.5) # Summarize significant genes
names(siggenes.table) # What data we have in the summary list
nrow(siggenes.table$genes.up) # How many upregulated genes
nrow(siggenes.table$genes.lo) # How many downregulated
# Or
siggenes.table$ngenes.up; siggenes.table$ngenes.lo
```

Let's have a look at actual differentially expressed genes:
```{r}
pander(head(siggenes.table$genes.up)) # Check how table with the results look like
```

We can write them in a clipboard (Windows only), then paste into Excel. Or export them to files:
```{r}
# Write the results in clipboard
# write.table(siggenes.table$genes.up,"clipboard-128",sep='\t')
# Write the results in file
write.table(siggenes.table$genes.up,"results/genes.up.txt",sep='\t') 
write.table(siggenes.table$genes.lo,"results/genes.dn.txt",sep='\t')
```

These results show pretty much everything we need to know. Except the Affy probe IDs do not tell any biological story up front - it would be nice to have at least gene names. Let's extract those IDs and annotate them:
```{r}
# Extract up- and downregulated IDs
up.ids<-siggenes.table$genes.up[,"Gene ID"] # Upregulated
dn.ids<-siggenes.table$genes.lo[,"Gene ID"] # Downregulated
```

We can simply get the platform data from GEO, and extract subsets of IDs:
```{r}
# Annotate Affy probe IDs
Meta(gds858)$platform # Check which platform do we have
gpl96<-getGEO('GPL96', destdir="data") # Get the data for this platform from GEO
pander(head(Table(gpl96)[, 9:12])) # Check the content of annotation data
pander(head(Table(gpl96)[Table(gpl96)[,"ID"] %in% up.ids,c("ID","Gene Symbol","Gene Title")])) # Extract annotation for up.ids
```

Or, better, we can use biomaRt. It is a more flexible approach:
```{r}
# Use biomaRt for annotation
biocLite("biomaRt") 
library(biomaRt)
pander(head(listDatasets(useMart("ensembl")))) # List available datasets
mart<-useMart("ensembl", dataset="hsapiens_gene_ensembl") # Load BIOMART dataset for homo sapiens
# Information - lists of filters and attributes
pander(head(listFilters(mart))) # Filters, these are our IDs we'll be subsetting the BIOMART annotations on
pander(head(listAttributes(mart))) # Attributes, these are annotations we would like to get for our IDs
Meta(gpl96)$title # Check which microarray we have, to select the right attributes
attr<-listAttributes(mart) # Get all attributes as a table
attr[grep("affy",attr[,1]),] # Parse them for anything that looks like from affymetrix
```

Now we know that our IDs correspond to "affy_hg_u133a" attributes in Biomart. Let's extract annotations for them:
```{r}
# Get annotations from Biomart
genes.up<-getBM(attributes=c('affy_hg_u133a','external_gene_name','description'), filters='affy_hg_u133a', values=up.ids, mart=mart)#, uniqueRows=T)
pander(head(genes.up)) # Check your work
```

Do that for the downregulated genes, and write the results to file.

Clean your workspace
```{r}
# Clean workspace
# unlink(c("GDS858.soft.gz","genes.up.txt","genes.dn.txt","GPL96.soft"))
```

Exercise 1
===
Using ALL dataset, find differentially expressed genes by limma and SAM. Start by importing and exploring the data.

Use pData(eset)$mol.bio as a factor for building matrix for limma, and for defining classes for SAM.




