---
dtitle: "Annotation_and_Enrichment_GENES"
author: "Mikhail Dozmorov"
date: "October 26, 2014"
output: html_document
---
```{r setup, echo=FALSE}
# Set up the environment
library(knitr) 
opts_chunk$set(cache.path='cache/', fig.path='img/', cache=T, tidy=T, fig.keep='high', echo=T, dpi=100, out.width=700)
options(replace.assign=TRUE, width=220)
options(stringsAsFactors = FALSE)
set.seed(1)
```

```{r}
library(DT)
library(ALL)
library(limma)
library(hgu95av2.db) # Load an annotation package for this platform
# biocLite("topGO")
library(topGO)
```

# Differential expression review with limma

Before proceeding into GOEA and other enrichment analyses, we need a list of probes and differential expression p-values. Here, we review limma usage with the ALL dataset. 

The ALL dataset has many covariates, which you can view with

```{r}
data(ALL) # Load actual dataset
datatable(pData(ALL))
colnames(pData(ALL)) # Which phenotype data we have
```

The ALL dataset contains either B-cell or T-cell leukemias, and  various cytogenetic abnormalities. Here, we compare expression in B-cell leukemia patients which have undergone BCR/ABL transition with B-cell leukemia patients which have no assigned molecular biology.

Probably 90% of all differential expression analyses have been done with two-group design, like "healthy controls" vs. "disease". Here, we are looking at a multifactorial design.

```{r}
pd <- pData(ALL)
# limma disallows special characters in factor names
mol.biol <- make.names(pd$mol.biol) # Convenient wrapper to make names compatible with R internals
groups <- factor(mol.biol, levels = unique(mol.biol))
design <- model.matrix(~0 + groups) # Multifactorial design matrix
colnames(design) <- levels(groups)
datatable(design) # Check your work
(contrast <- makeContrasts(BCR.ABLvsNEG = BCR.ABL-NEG, levels=design)) # Define what to compare with what
M <- normalizeQuantiles(exprs(ALL)) # Normalize gene expression matrix before differential expression
fit <- lmFit(M, design)
fit.c <- contrasts.fit(fit, contrast)
fit.c <- eBayes(fit.c)
tt <- topTable(fit.c, number=nrow(exprs(ALL)), coef = "BCR.ABLvsNEG", adjust.method = "BH", p.value = 0.05) # All results of differential expression analysis
# For the enrichment analysis, we extract a vector of adjusted p-values, named by probe IDs
p.values <- tt$adj.P.Val
names(p.values) <- rownames(tt)
datatable(as.matrix(p.values))
```

Optionally, we check which gene NAMEs are differentially expressed

```{r}
result <- decideTests(fit.c) # Correct the results for multiple testing
sig.probes <- rownames(result)[result!=0] # Pull up the probe names 

ALL@annotation # Check which microarray platform was used. Or, annotattion(ALL)
xx <- as.list(hgu95av2SYMBOL) # Get full mapping table for probes to gene names
sig.symbols <- unlist(xx[sig.probes]) # Extract gene names from the significant probes
head(sig.symbols) # Check your work
length(unique(sig.symbols)) # How many significant genes are there?
```

# Gene Ontology Enrichment Analysis

The *topGO* package is a gold standard for microarray differential expression analysis. We need to create a topGOdata object from pieces we obtained from limma analysis.

```{r}
sel.fn <- function(p.vals) { p.vals < 0.01 } # Returns a subset of most significant probes
affyLib <- paste(annotation(ALL), "db", sep=".")
go.data <- new("topGOdata",
               ontology = "BP", allGenes = p.values, geneSel = sel.fn,
               nodeSize = 10, # search GO terms with >= 10 genes
               annot = annFUN.db, affyLib = 'hgu95av2.db')
```

A variety of algorithms and ranking statistics are available for running the actual enrichment step. A list of algorithms and tests can be found by calling:

```{r}
whichAlgorithms()
whichTests()
```

We try a couple:

```{r, message=FALSE}
result <- runTest(go.data, algorithm="classic", statistic="fisher")
result.01 <- runTest(go.data, algorithm="weight01", statistic="fisher")
result
```

View the top results from multiple methods in tabular format

```{r}
datatable(GenTable(go.data, result, result.01))
```

Let's have a look at the results in the context of the GO DAG.

```{r}
showSigOfNodes(go.data, score(result), firstSigNodes = 3, useInfo = "all")
```

Very small text. This can be easier to read if you save into an external file with pdf() and dev.off(). 


# Creating a custom enrichment analysis

At its core, enrichment analysis is very simple. You have a list of genes, a list of categories, and a list of gene-category associations.

The question is, given a list of genes (for example, the top differentially expressed genes from an experiment), what categories are these genes statistically enriched for? The standard test for this is Fisher Exact.

This simplicity means enrichment analysis is very broadly applicable. Any time you have a data source that annotates genes with categories, you can use enrichment analysis.

Back to our example with genes enrched in angiogenesis.

* Out of 12,350 genes, 185 are involved in “Angiogenesis”. 
* In our experiment, out of 200 differentially expressed genes 35 are annotated with “Angiogenesis” function

```{r}
# Fisher Exact test with contingency table
ct <- matrix(c(35, 165, 150, 12000), byrow = T, nrow = 2)
ct
fisher.test(ct, alternative = "greater")
```

We can wrap it up into a "universal" function that does either GO or KEGG enrichment analysis. see the example at [https://github.com/mdozmorov/MDmisc/blob/master/R/gene_enrichment.R](https://github.com/mdozmorov/MDmisc/blob/master/R/gene_enrichment.R)


# Exercise

Write a function that takes a vector of significantly  differentially expressed probes and an annotation name, and uses Fisher's Exact test to test whether any chromosomes are significantly enriched or depleted for differentially expressed genes. Be sure to correct for multiple testing.

To do this, you might find the "get" function useful. For example:

```
get(paste("hgu95av2", "CHRLOC", sep=""))
```

returns the CHRLOC map for the hgu95av2 annotation package, assuming it has been loaded already by library().
